Editor: Display spinner while featured image loads

aduth opened the pull request:
Related: #4510

This pull request seeks to enhance the featured image component to display a spinner while the featured image loads. Currently, when loading the editor, once a featured image is detected to exist, the header borders are shown, but the image does not display until it has loaded, at which point it appears suddenly. Since we can know earlier in the page lifecycle that a featured image exists, we can display a loading indicator while the image is being shown. This will also improve the feeling of responsiveness in #4510 when changing existing featured images.

Before|After
---|---
![before](https://cloud.githubusercontent.com/assets/1779930/14475129/65700488-00ce-11e6-9096-956b7b57012a.gif)|![after](https://cloud.githubusercontent.com/assets/1779930/14475128/656f6a6e-00ce-11e6-8c4d-996744b0eeb5.gif)

__Design Notes:__

Per previous discussion, it's argued that the use of a spinner is justified in indicating an incoming image:

>Premise 1: Absence of any loading indicator may lead user to believe that no preview will be rendered, and provides a jarring experience once the image loads.
>Premise 2: Using the traditional "pulsing" element effect may wrongly convince the user that the placeholder itself is the previewed item
>Conclusion: <img src="https://cldup.com/JiXZUOcvQM.svg" width="16" height="16" />

Ref: 8565-gh-calypso-pre-oss

__Testing instructions:__

Verify that no regressions occur in the display of a featured image, whether its in the header of in the sidebar. Display of a featured image should mirror that of the master branch. Notably, when displayed in the header, an image should be restricted to a maximum height of 400 pixels.

~~Note that the included changes try to adjust the height of the container to the known height of the image prior to its being loaded, but because the media must first be loaded, we display the container at 400 pixels tall until the height of the image is known. Therefore, for images smaller than 400 pixels tall, it'll appear briefly as taller, then shrink to the appropriate size.~~ Due to complications noted in https://github.com/Automattic/wp-calypso/pull/4687#issuecomment-216995101 , height estimating behavior has been removed.

1. Navigate to the [post editor](http://calypso.localhost:3000/post)
2. Select a site, if prompted
3. Expand the Featured Image accordion
4. Click Set Featured Image
5. Choose and assign a featured image
6. Note that the featured image displays correctly in the header and sidebar
7. Enter a title or post content
8. Save the post
9. Refresh the page
10 Note that the following progression occurs while the page loads:
 - Until the post is loaded, no featured image element is shown on the page
 - When the post loads and a featured image exists, until the media item has been loaded, a placeholder will be shown in the header
 - ~~When the featured image media item is loaded but the image itself has yet to be loaded, the placeholder should occupy the space equivalent to its height, capped at 400 pixels tall~~
 - When the featured image is shown, it replaces the placeholder and the spinner is no longer shown

folletto commented:
>  Therefore, for images smaller than 400 pixels tall, it'll appear briefly as taller, then shrink to the appropriate size.

My advice here would be to try to always grow in one direction: from zero to the minimum amount to show the spinner to the real final height.

aduth commented:
>My advice here would be to try to always grow in one direction: from zero to the minimum amount to show the spinner to the real final height.

In retrospect, I'd agree with this. One problem we may encounter is there's no real minimum for a featured image height, so how would we expect the height to grow in the case of a very short image? Still display the minimum for the spinner (with some padding) and accept that it's less common that the featured image would be shorter than that?

folletto commented:
I'd say so. I feel that "shorter than the spinner" it's very unlikely to be a featured image. :)

folletto commented:
I just had an idea. :)

![cap-](https://cloud.githubusercontent.com/assets/4389/14605280/36b609e0-0578-11e6-9fe9-ca719f57d6dc.gif)

Here's the live example with the CSS to borrow from: http://codepen.io/folletto/full/JXvxRb/

aduth commented:
@folletto :

First thought: Spinner that doesn't look like a spinner for the sake of not looking like a spinner 😆 

Second thought: Maybe you're on to something, specifically as a pattern for indicating a "to-be-expanded-when-loaded" section pending state. I can buy it 👍

jasmussen commented:
I like the thin line primarily because of the pattern you mention, Andrew:

> a pattern for indicating a "to-be-expanded-when-loaded" section pending state

A scourge of the web is unloaded images that expand and makes the page jump as they load and their height becomes available. However in absence of eliminating that jump, I have a preference for things that expand vertically, as opposed to collapsing down. So that's basically echoing what Davide says here:

> always grow in one direction: from zero to the minimum amount to show the spinner to the real final height

For those reasons both I like the thin bar. In addition, if we were to add more media types that you can insert, and more features around inserting those, the thin-bar-pattern might come in handy as an indicator of where you'll be dropping a media item — think dragging and dropping the bar between paragraphs, for it to then expand to the media item.

folletto commented:
Alternative coloring option using grays instead of blues:

```background-image: linear-gradient( 90deg, #a8bece 0%, #c8d7e1 50%, #a8bece 100% );```

![cap-](https://cloud.githubusercontent.com/assets/4389/14851305/e50bdbb4-0c76-11e6-8ef6-e9c2f4d0c7f1.gif)

aduth commented:
I'm hoping to circle back around to this (and #4510) this week with the proposed improvements.

aduth commented:
We'll also need to consider how the pending state should be displayed for the Featured Image sidebar accordion.

folletto commented:
> We'll also need to consider how the pending state should be displayed for the Featured Image sidebar accordion.

True. 

I think we can keep the original size, and just add the standard spinner in there. :)

![featured-image-sidebar-placeholder-i1](https://cloud.githubusercontent.com/assets/4389/14985367/164217f8-113f-11e6-999a-8d7a0c008abd.png)

aduth commented:
Iterated on this pull request more this afternoon. I've included a newly introduced `<SpinnerLine />` component per the discussion here, which will be shown while the post loads or is transitioning to a new assigned feature image (not possible until #4510).

![load](https://cloud.githubusercontent.com/assets/1779930/15028185/9491337e-1215-11e6-92cb-1df1decb4c62.gif)

Changes include:
- Removed logic around trying to assume a specific height for the image when the data is known but the image resource itself hadn't been loaded. There are complications around site thumbnail sizes and we currently don't have accurate means of estimating the correct height.
- Display `<SpinnerLine />` instead of `<Spinner />` while image loads
- Preserve height of container when switching images so that height doesn't collapse and expand sporadically while image loads
- To preserve current behavior of gracefully transitioning from an in-progress upload to its uploaded copy, added new checks when receiving a new image to check whether it's the persisted copy of the previous image, showing the previous instead of a spinner while the persisted copy loads.

Not included here is the revisions from the comment above https://github.com/Automattic/wp-calypso/pull/4687#issuecomment-216535449 . Due to component re-use between the header and sidebar featured image components, there's not a graceful solution to showing different variations of spinners between the two uses. I think the behavior present in these changes (spinner line) is an improvement over the master branch (zero height with "remove" button present), at least enough to hold us over 'til a subsequent pull request.

timmyc commented on client/components/spinner-line/README.md:
```diff
@@ -0,0 +1,32 @@
+
+<img src="https://cldup.com/cYmgv_L6r3.gif" alt="Demo" />
+
+__Please exercise caution in deciding to use a spinner in your component.__ A lone spinner is a poor user-experience and conveys little context to what the user should expect from the page. Refer to [the _Reactivity and Loading States_ guide](https://github.com/Automattic/wp-calypso/blob/master/docs/reactivity.md) for more information on building fast interfaces and making the most of data alread ...
```
👍

timmyc commented:
Code changes are looking good to me, and the new line spinner tests out well too.  Before marking ready to merge I figure some of the others on this PR should give the new component a design review.

folletto commented:
Tested in Chrome, works well.
Ok on the consideration regarding component reuse.

If possible, I'd just make a style adjustment:

![screen shot 2016-05-05 at 13 53 38](https://cloud.githubusercontent.com/assets/4389/15043917/f0044bb6-12c8-11e6-8f99-4972905d8b59.png)

In other terms, avoid showing the inset area (`.editor-featured-image`) immediately, just show the `<SpinnerLine>` and then when the image loads open it within the inset area.

Doable?

aduth commented:
>Doable?

Yep, should be easy enough, and makes sense to me.

aduth commented:
>Doable?

Updated in b07a342.

folletto commented:
Tested in Chrome, cool! 👍  :shipit:
